python/shotgun_desktop/startup.py

```diff
@@ -698,6 +699,45 @@ def get_bundle_cache_location(self):
             return os.environ.get("SHOTGUN_DESKTOP_BUNDLE_CACHE_LOCATION") or None
 
 
+def automatic_claims_renewal(user, preemtive_renewal_threshold=0.9):
+    """
+    Handles automatic renewal of the SAML2 claims for the user.
+
+    :params user: an already logged in user.
+    :params preemtive_renewal_threshold: How far into the claims duration we will attempt renewal.
+                                         Defaults to 90%, usually 3 minutes 45 seconds.
+
+    This will only be used in a context where SSO is enabled on the
+    Shotgun site.
+    """
+    from tank.authentication import interactive_authentication
+    from sgtk.authentication import AuthenticationCancelled
+
+    logger.debug("Automatic claims renewal")
+    try:
+        previous_expiration = user._impl.get_saml_expiration()
```

Comment:
My logic was this : the background renewal of the claims is purely done as a convenience feature for the user of the Shotgun Desktop (or any long-running application, e.g. more than 5 mins). Otherwise, after an hour of inactivity (actually, this is determined at IdP level and there is now way to know what that on the client side) the user would need to re-enter their credentials... for example coming back from lunch.

Given that this was an application need, I believed that it should be dealt with at the application level. Not in the core.

The call to renew_session() is safe to do at the desktop level... We need to nicely expose a) the fact that SSO is on, and b) obtain the expiration of the saml claims.